aten/src/ATen/native/mps/operations/LinearAlgebra.mm

```diff
@@ -91,6 +94,114 @@ void prepare_matrices_for_broadcasting(const Tensor* bias,
 
 enum LinearAlgebraOpType { ADDBMM_OP_TYPE, BADDBMM_OP_TYPE };
 
+void linalg_lu_factor_out_mps_impl(const Tensor& A, bool pivot, Tensor& LU, Tensor& pivots) {
+  using namespace mps;
+
+  Tensor A_t = A;
+  uint64_t aRows = A_t.size(-2);
+  uint64_t aCols = A_t.size(-1);
+  uint64_t aElemSize = A_t.element_size();
+  uint64_t numPivots = std::min(aRows, aCols);
+  resize_output(LU, A_t.sizes());
+  std::vector<int64_t> pivot_sizes(A_t.sizes().begin(), A_t.sizes().end() - 2);
+  pivot_sizes.push_back(numPivots);
+  resize_output(pivots, pivot_sizes);
+
+  if (A_t.numel() == 0 || LU.numel() == 0) {
+    LU.zero_();
+    return;
+  }
+
+  Tensor A_ = A_t.dim() > 3 ? A_t.flatten(0, -3) : A_t;
```

Comment:
`flatten` may copy. Do you really need to copy here? You can see how this is done in CPU/CUDA for inspiration on how to not perform unnecessary copies.

Reply:
We don't need copy here but end up need `A_` to be contiguous before feeding it into the kernel. Can you please point out where it is done in CPU/CUDA?

Reply:
Hi @lezcano, can you give me some pointer here on how the cpu/cuda avoid unnecessary copies?

Reply:
Used view instead